File: hsu-auto-scheduler-fe/src/types/schemas/Course.schema.ts
import { DayOrNightEnum } from "@/enums/dayOrNight.enum";
import z from "zod";
import { offlineScheduleSchema } from "./OfflineSchedule.schema";

export const courseSchema = z.object({
  semester_id: z.string().min(1, { message: "학기를 선택해주세요!" }),

  course_id: z.string().min(1, { message: "course id는 필수값입니다" }),

  course_code: z.string().min(1, { message: "강의 코드는 필수값입니다" }),

  course_name: z.string().min(1, { message: "강의명은 필수값입니다" }),

  professor_names: z.array(
    z.string().min(1, { message: "professor name은 문자열이어야 합니다" }),
  ),
  completion_types: z.array(
    z.string().min(1, { message: "이수 구분은 필수값입니다" }),
    { message: "이수 구분은 배열이어야 합니다" },
  ),

  delivery_method: z.string().min(1, { message: "과목 구분은 필수값입니다" }),

  credit: z
    .number({ message: "학점은 숫자여야 합니다" })
    .min(1, { message: "학점의 최솟값은 1이상이어야 합니다" }),

  day_or_night: z.enum(DayOrNightEnum),

  class_section: z
    .string({ message: "분반은 문자열이어야 합니다" })
    .min(1, { message: "분반을 입력해주세요" }),

  grades: z.array(z.number({ message: "grade는 숫자여야 합니다." }), {
    message: "grades는 숫자의 배열이어야 합니다",
  }),

  grade_limit: z.union([z.number(), z.null()]).optional(),

  online_hour: z.number({
    message: "온라인 시간은 숫자여야 합니다.",
  }),

  offline_schedules: z.array(offlineScheduleSchema),

  plan_code: z.union([z.number(), z.null()]).optional(),
});

export type CourseType = z.infer<typeof courseSchema>;

File: ortools-scheduler/app/schemas/common/course_schema.py
from pydantic import BaseModel, Field
from typing import Optional
from app.schemas.common.enums import DayOrNightEnum
from app.schemas.common.offline_schedule_schema import OfflineScheduleSchema


class CourseSchema(BaseModel):
    semester_id: str = Field(..., description="학기 ID")
    course_id: str = Field(..., description="강의 ID")
    course_code: str = Field(..., description="강의 코드")
    course_name: str = Field(..., description="강의 이름")
    professor_names: list[str] = Field(..., min_items=1, description="교수 이름 리스트")
    completion_types: list[str] = Field(..., description="이수 구분")
    delivery_method: str = Field(..., description="수업 방식")

    credit: int = Field(..., description="학점", ge=0)
    day_or_night: DayOrNightEnum = Field(..., description="주간/야간 여부")

    class_section: str = Field(..., description="분반")
    grades: list[int] = Field(..., description="학년")

    grade_limit: Optional[int] = Field(None, description="학년 제한")
    online_hour: float = Field(..., description="온라인 수업 시간", ge=0)
    offline_schedules: list[OfflineScheduleSchema] = Field(
        None, description="오프라인 시간표"
    )
    plan_code: Optional[str] = Field(None, description="계획 코드")
